content/blog/2019-11-13-gep.md/index.md

```diff
@@ -0,0 +1,458 @@
+```
+### LLVM Address Calculation
+
+When compiling high level array and struct access to LLVM code, compilers
+generally use the [getelementptr](https://llvm.org/docs/GetElementPtr.html)
+(or GEP) instruction to calculate offsets into these memory allocations.
+GEP instructions have the nice property that they are type aware; offsets are phrased
+in terms of "number of elements" rather than "number of bytes."
+For example, the code in this stub dereferences some memory in
+the middle of a struct (specifically the last element of the `b` field).
+
+```C
+struct EX {
+  int a;
+  char b[3];
+  int *c;
+}
+
+struct X;
+...
+return X->b[2];
+```
+
+In LLVM, we could write a single GEP instruction to calculate
+the correct offset into the struct and then execute a `load` instruction
+to actually dereference the pointer.
+
+```C
+%1 = getelementptr %struct.EX,  %struct.EX* %X, i64 0, i32 1, i64 2
+%2 = load i8, i8* %1
+ret %2
+```
+
+This functionality makes GEP an ideal point for analyzing out-of-bounds accesses.
+Before a program might make an out-of-bounds access it has to
+acquire an out-of-bounds pointer. Usually, this means it
+executes a GEP whose result will then later be the argument of a
+load or store operation.
+
+Our approach in this implementation is to prevent the execution of
+any runtime GEP instructions that might lead to
+illegal memory accesses.
+*If a program can never acquire an out-of-bounds pointer,
+it can't violate memory safety.*
+
+### Making GEP Safe(r)
+
+Let's go back to our first example of an out of bounds array access:
+
+```C
+int tmp[10];
+...
+return tmp[x];
+```
+
+The return statement roughly translates to:
+
+```C
+%addr = getelementptr [10 x i32], [10 x i32]* %tmp, i64 0, i64 %x;
+%val = load i32, i32* %addr;
+ret %addr;
+```
+
+In order to insert a dynamic check for memory safety, there
+are two things we need to know:
+
+ - What is the *actual* access index value?
+ - What are **legal** access index values?
+
+Happily, when considering GEP instructions, the first question is easy
+to answer; each operand represents an access index
+value. We can dynamically insert instructions into the program that compare
+those operands to other values.
+
+The second question is a much more difficult problem,
+whose subtleties we'll address in the next section.
+For the most part though, we leverage LLVM's type
+information. Based on its type annotation, we know
+ that `%tmp` points to an array with 10 32-bit integers (notated as `[10 x i32]`).
+Therefore, we can conclude that the only valid values for `%x` are
+between 0 (inclusive) and 10 (exclusive).
+
+To execute this check completely, we need to check that *all* index operands
+are legal. You may have noticed that our prior example actually has **two**
+index operands; we have to check both that `tmp` points to at least one
+integer array of size 10, *and* that `x` is a valid index for such an array.
+
+In general our algorithm for modifying LLVM code is this:
+
+```
```

Comment:
Probably no need for a code block here—you can use a nice Markdown numbered list, like this:

1. Foo.
2. Bar.